crates/wasi/src/filesystem.rs

```diff
@@ -236,45 +252,175 @@ impl Dir {
+                self.state = match task.await {
+                    Ok(b) if b.len() == 0 => ReadState::Closed,
+                    Ok(b) => ReadState::DataAvailable(b),
+                    Err(e) => ReadState::Error(e),
+                };
+            }
+            _ => {}
+        }
+    }
+
+    fn read_available_data(&mut self, size: usize) -> Bytes {
+        let ReadState::DataAvailable(b) = &mut self.state else {
+            unreachable!();
+        };
+        debug_assert!(b.len() > 0);
+
+        if size == 0 {
+            return Bytes::new();
+        }
+
+        let min_len = b.len().min(size);
+        let chunk = b.split_to(min_len);
+        if b.len() == 0 {
+            self.state = ReadState::Idle;
+        }
+        self.position += min_len as u64;
+        chunk
+    }
+}
+#[async_trait::async_trait]
+impl HostInputStream for FileInputStream {
+    fn read(&mut self, size: usize) -> StreamResult<Bytes> {
+        match self.state {
+            ReadState::Idle => {
+                if size == 0 {
+                    return Ok(Bytes::new());
+                }
+
+                let p = self.position;
+                self.state = ReadState::Waiting(
+                    self.file
+                        .spawn_blocking(move |f| Self::blocking_read(f, p, size)),
+                );
+                Ok(Bytes::new())
+            }
+            ReadState::DataAvailable(_) => Ok(self.read_available_data(size)),
+            ReadState::Waiting(_) => Ok(Bytes::new()),
+            ReadState::Error(_) => match mem::replace(&mut self.state, ReadState::Closed) {
+                ReadState::Error(e) => Err(StreamError::LastOperationFailed(e.into())),
+                _ => unreachable!(),
+            },
+            ReadState::Closed => Err(StreamError::Closed),
+        }
+    }
+    /// Specialized blocking_* variant to bypass tokio's task spawning & joining
+    /// overhead on synchronous file I/O.
+    async fn blocking_read(&mut self, size: usize) -> StreamResult<Bytes> {
+        self.wait_ready().await;
+
+        match &mut self.state {
+            ReadState::Idle => {
+                if size == 0 {
+                    return Ok(Bytes::new());
+                }
+
+                let p = self.position;
+                match self
+                    .file
+                    .run_blocking(move |f| Self::blocking_read(f, p, size))
+                    .await
+                {
+                    Ok(b) if b.len() == 0 => {
+                        self.state = ReadState::Closed;
+                        Err(StreamError::Closed)
+                    }
+                    Ok(b) => {
+                        self.position += b.len() as u64;
+                        Ok(b)
+                    }
+                    Err(e) => {
+                        self.state = ReadState::Closed;
+                        Err(StreamError::LastOperationFailed(e.into()))
+                    }
+                }
```

Comment:
The implementation of this seems a little subtle and while I see no issue with it I think it would be best to share as much as possible with the above code. Could something like this work?

* In `blocking_read` first resolve the `Waiting` state if any with a `.await`
* Next handle the `Idle` state by performing a `run_blocking` but the result of that is handled in a shared way with what `wait_ready` does (setting `self.state` internall)
* Next handle the results of the read in the same way as `fn read` above (perhaps even just calling `self.read(size)`?)

That I think will keep the `blocking_read` bits here deduplicated with the rest and the only responsibility of `blocking_read` vs `read` would be to filter the states that `read` can see by doing blocking work up front.

Reply:
Something like [this](https://github.com/bytecodealliance/wasmtime/pull/9129/commits/7e2b7e9ec507313d13f877d6f7a1557577a2578a)?

Reply:
Perfect!